bpf/bpf_overlay.c

```diff
@@ -75,13 +75,15 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
 
 	decrypted = ((ctx->mark & MARK_MAGIC_HOST_MASK) == MARK_MAGIC_DECRYPT);
 	if (decrypted) {
-		if (info)
-			*identity = key.tunnel_id = info->sec_identity;
+		if (info) {
+			*identity = info->sec_identity;
+			key.tunnel_id = get_tunnel_id(info->sec_identity);
```

Comment:
I don't think that we need to change this case since the security identity is coming from ipcache?

Even if we need to change this line, I don't think that `*identity` and `key.tunnel_id` should diverge.

Reply:
An ipcache lookup could still result in the identity being `WORLD_IPV6_ID`.

Reply:
That's correct on receive, right? Internally the datapath would understand `WORLD_IPV6_ID`, it's only when we serialize/deserialize we need to understand that peers may be transmitting `WORLD_ID`.

Reply:
^ Reviewing afresh with Nathan,this path is updating `key.tunnel_id` to ensure that whenever we modify that field, we always encode `WORLD_IPV*_ID` into `WORLD_ID`, and whenever decoding from that field, we convert back.

I'm still a little bit concerned about the potential for divergence here, but this is probably easier to ensure that the implementation is correct because any subsequent reads from the `tunnel_id` field will convert back. I withdraw my comment here :+1: